### PR TITLE
Record: 11L XSA-all + Full GPTQ (Budget-Legal) + Parallel Muon + Selective Pruning (val_bpb: 1.1178, 3-seed mean)

### PR DESCRIPTION
**Pull request by raahilshah**
# Record: 11L XSA-all + Full GPTQ (Budget-Legal) + Parallel Muon + Selective Pruning

**val_bpb: 1.1178** (3-seed mean, std 0.0001) | **15.95 MB** max artifact | 8xH100 SXM, ~596s total compute

## Update (2026-03-26)

This PR was updated to fix a GPTQ budget violation identified in [issue #677](https://github.com/openai/parameter-golf/issues/677). The previous version trained for the full 600s, then ran GPTQ calibration for ~46s on top. The fix reserves 14s from the training budget (`gptq_reserve_ms = 14000.0`), ensuring total compute stays within 600s. See [update comment](https://github.com/openai/parameter-golf/pull/634#issuecomment-4136949992) for details.

## Results (3 seeds, 8xH100 SXM)

| Seed | Sliding BPB (s64) | val_loss | Artifact | Train Time | GPTQ Time | Total |
|------|--------------------|----------|----------|------------|-----------|-------|
| 1337 | **1.1177** | 1.8871 | 15,929,433 B | 586,128ms | 9,786ms | 595,915ms |
| 42 | 1.1179 | 1.8875 | 15,949,353 B | 586,050ms | 9,792ms | 595,842ms |
| 7 | 1.1179 | 1.8875 | 15,946,145 B | 586,066ms | 9,823ms | 595,889ms |

**Mean: 1.1178 | Std: 0.0001**

## Key Techniques

1. **XSA on all 11 layers** — cross-position mixing from layer 0 (-0.0016 BPB vs XSA-4)
2. **Full Hessian GPTQ** — 64-batch GPU Hessian, Cholesky error compensation, **budget-legal** (14s reserved)
3. **amax-aligned QAT** — STE matches export quantizer, [-32, 31] range
4. **Parallel Muon** with parameter banking — 3-phase overlapped optimizer, ~87ms/step
5. **Selective magnitude pruning** — post-GPTQ, zero least-impactful ±1 values
6. **LZMA compression** — preset 6, better than zstd on int6 weights
7. **LeakyReLU(0.5)²** — prevents dead neurons

## Architecture

11L, 512d, 8H/4KV (GQA), MLP 3x LeakyReLU(0.5)², XSA-all-11, Partial RoPE 16/64, LN Scale, VE128, SmearGate, BigramHash(2048), U-Net skips, EMA(0.997), Tight SWA, Late QAT@0.15, Full GPTQ int6 + LZMA, Parallel Muon + Parameter Banking, FA3 Hopper.

## Compliance

- [x] 3 seeds, all total compute ≤600s (max: 595,915ms, verified via `gptq:budget_check` in logs)
- [x] GPTQ calibration WITHIN training budget (14s reserved from 600s)
- [x] All artifacts ≤16,000,000 bytes (max: 15,949,353)
- [x] No TTT on validation data
- [x] No training data accessed during evaluation
- [x] No network calls during evaluation
- [x] Sliding window eval stride=64 (std=0.0001)

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Record: 11L XSA-all + Full GPTQ (Budget-Legal) + Parallel Muon + Selective Pruning (val_bpb: 1.1178, 3-seed mean)

**BPB:** 1.1178 | **Compliance:** FLAG — hashed n-gram cache with target-in-key (PR #779 family pattern)

**What I found in the code** (head SHA `864d15389c66`, file `records/track_10min_16mb/2026-03-24_11L_XSA-all_GPTQ_ParallelMuon_1.1171/train_gpt.py`):

The n-gram lookup key at line **1319** is constructed by XOR-ing the target token into the hash:

```
line 1319: full_key = <hash> ^ (tgt_arr * ng_primes[...]) & mask
```

This matches the `full_key = ((ctx_hash ^ (target * primes[k])) & mask)` construction that @valerio-oai ruled disallowed on PR #779 ([comment 4145781641](https://github.com/openai/parameter-golf/pull/779#issuecomment-4145781641), 2026-03-27). Per the [mechanism explanation](https://github.com/openai/parameter-golf/pull/779#issuecomment-4146407380), hashing the target token into the lookup key only reweights the correct token — in the hash-collision limit this drives P(correct) → 1 regardless of the data, which inflates the reported BPB without producing real compression.

Per **Issue #1017 condition 1**, `p_t` may depend only on the artifact and `x_1...x_{t-1}`. Because the lookup key at line 1319 is a function of the target token, the count read at scoring position *t* depends on *x_t* itself — which is the core violation the #779 ruling targets.

**Cluster context:** this same structural pattern has been closed on 15+ PRs under the #779 ruling as of 2026-04-11 (#779 itself, #770, #798, #808, #825, #786, #797, #909, #940, #761, #776, #788, #774, #778, #715, #758, #702 upstream, #1488). The base neural model is unaffected by this flag — in every case where the authors resubmitted without the n-gram cache, the base val_bpb has been in the ~1.10-1.15 range (standard for the SP1024 11L class).

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 0.06s, dim=512, layers=11, vocab=1024, code=113761 B, SMOKE_TEST_PASS

**Verdict:** COMPLIANCE FLAG — target-in-key hashed n-gram cache, same family as PR #779.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **CLOSE** under the same ruling as the rest of the family-bug cluster. A context-only resubmission (drop the target from the lookup key and use a full-vocabulary reweighting from a single context row, per @valerio-oai's suggested legal path on #779) would be welcomed.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 0.06s, dim=512, layers=11, vocab=1024, code=113761 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*
